eng/tools/spec-gen-sdk-runner/src/utils.ts

```diff
@@ -94,5 +132,289 @@ export function runPowerShellScript(args: string[]): string | undefined {
+      if (files.some((file) => searchFile.test(file.toLowerCase()))) {
+        return currentPath;
+      }
+    } catch (error) {
+      logMessage(`Error reading directory: ${currentPath} with ${error}`, LogLevel.Warn);
+      return undefined;
+    }
+    currentPath = path.dirname(currentPath);
+    if (stopAtFolder && currentPath === stopAtFolder) {
+      return undefined;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Searches for parent directories containing specific files for a list of files
+ * Optimizes the search by grouping files in the same directory to avoid redundant searches
+ * @param files - Array of file paths to process
+ * @param options - Search configuration options
+ * @returns Object mapping parent directory paths to arrays of related files
+ */
+export function searchRelatedParentFolders(
+  files: string[],
+  options: { searchFileRegex: RegExp; specRepoFolder: string; stopAtFolder?: string },
+): { [folderPath: string]: string[] } {
+  const result: { [folderPath: string]: string[] } = {};
+
+  // Group files by their directory path to avoid redundant searches
+  // Example: for files ["dir1/a.ts", "dir1/b.ts", "dir2/c.ts"]
+  // Creates: { "dir1": ["dir1/a.ts", "dir1/b.ts"], "dir2": ["dir2/c.ts"] }
+  // eslint-disable-next-line unicorn/no-array-reduce
+  const filesByDir = files.reduce<{ [dir: string]: string[] }>((acc, file) => {
+    const dir = path.dirname(file);
+    if (!acc[dir]) {
+      acc[dir] = [];
+    }
+    acc[dir].push(file);
+    return acc;
+  }, {});
+
+  // Search parent folder only once per unique directory
+  for (const [dir, dirFiles] of Object.entries(filesByDir)) {
+    const parentFolder = findParentWithFile(
+      dir,
+      options.searchFileRegex,
+      options.specRepoFolder,
+      options.stopAtFolder,
+    );
+    if (parentFolder) {
+      if (!result[parentFolder]) {
+        result[parentFolder] = [];
+      }
+      result[parentFolder].push(...dirFiles);
+    }
+  }
+
+  return result;
+}
+
+/**
+ * Identifies files that are part of a shared library based on their directory names
+ * @param files - Array of file paths to check
+ * @param options - Search configuration options
+ * @returns Array of files that belong to shared libraries
+ */
+export function searchSharedLibrary(
+  files: string[],
+  options: { searchFileRegex: RegExp; specRepoFolder: string },
+): string[] {
+  return files.filter((file) => {
+    const dirname = path.dirname(file);
+    return options.searchFileRegex.test(dirname);
+  });
+}
+
+/**
+ * Finds peer TypeSpec projects for shared libraries and maps them to their source libraries
+ * Assumes all shared libraries are from the same parent folder
+ * @param sharedLibraries - Array of shared library file paths (all from same parent)
+ * @param options - Search configuration options
+ * @returns Object mapping project directory paths to arrays of related shared library files
+ */
+export function searchRelatedTypeSpecProjectBySharedLibrary(
```

Comment:
Just as an FYI if we end up taking on Jeff directory structure changes there will not be a notion of a shared library.

Reply:
That's right. Regardless of the final version of the new folder layout, our tools will need to be updated to support the changes. This version is to support the current folder layout.